pkg/daemon/ceph/osd/volume.go

```diff
@@ -460,7 +493,89 @@ func UpdateLVMConfig(context *clusterd.Context, onPVC, lvBackedPV bool) error {
 	return nil
 }
 
-func (a *OsdAgent) initializeDevices(context *clusterd.Context, devices *DeviceOsdMapping) error {
+func (a *OsdAgent) useRawMode(context *clusterd.Context, pvcBacked bool) (bool, error) {
+	if pvcBacked {
+		return a.clusterInfo.CephVersion.IsAtLeast(cephVolumeRawModeMinCephVersion), nil
+	}
+
+	var useRawMode bool
+	// Can we safely use ceph-volume raw mode in the non-PVC case?
+	// On non-PVC we see a race between systemd-udev and the osd process to acquire the lock on the device
+	if a.clusterInfo.CephVersion.IsNautilus() && a.clusterInfo.CephVersion.IsAtLeast(cephFlockFixNautilusMinCephVersion) {
+		logger.Debugf("will use raw mode since cluster version is at least %v", cephFlockFixNautilusMinCephVersion)
+		useRawMode = true
+	}
+
+	if a.clusterInfo.CephVersion.IsOctopus() && a.clusterInfo.CephVersion.IsAtLeast(cephFlockFixOctopusMinCephVersion) {
+		logger.Debugf("will use raw mode since cluster version is at least %v", cephFlockFixOctopusMinCephVersion)
+		useRawMode = true
+	}
+
+	// ceph-volume raw mode does not support encryption yet
+	if a.storeConfig.EncryptedDevice {
+		logger.Debug("won't use raw mode since encryption is enabled")
+		useRawMode = false
+	}
+
+	// ceph-volume raw mode does not support more than one OSD per disk
+	osdsPerDeviceCountString := sanitizeOSDsPerDevice(a.storeConfig.OSDsPerDevice)
+	osdsPerDeviceCount, err := strconv.Atoi(osdsPerDeviceCountString)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to convert string %q to integer", osdsPerDeviceCountString)
```

Comment:
Nit.

The error returned from `Atoi` is
```
strconv.Atoi: parsing "<some string>": invalid syntax
```

I don't think we need to note the string in our error. It might be best to include more context like why the string-int conversion is needed.